### PR TITLE
tighten pod status validation

### PR DESCRIPTION
**Pull request by deads2k**
ContainerStateValidation allows some transitions that I don't think are intended.  I've tightened a couple here and inserted some comments asking for clarification about what is and is not intentional

@sjenning ptal and comment.  I'm not attached to this, but I'd like to codify what we think should happen.

/kind bug
/priority important-longterm
@kubernetes/sig-node-pr-reviews

### CODE REVIEW

**Comment by deads2k** [comment]
/hold

this is to gather comments before I add unit tests.

File: pkg/apis/core/validation/validation.go [hunk 1]
```diff
@@ -3570,9 +3586,22 @@ func ValidateContainerStateTransition(newStatuses, oldStatuses []core.ContainerS
...
+			if newStatus.State.Waiting != nil {
+				allErrs = append(allErrs, field.Forbidden(fldpath.Index(i).Child("state"), "may not go back to waiting"))
+			}
+			if newStatus.State.Running != nil {
+				allErrs = append(allErrs, field.Forbidden(fldpath.Index(i).Child("state"), "may not go back to running"))
+			}
+
+			// TODO this allows us to set Image, ImageID, and ContainerID back to empty after they had a value.  Is this intentional?  Should it accompanied by a state change?
+			// TODO this allows restartCount to move backwards.  Is this intentional?
+			// TODO this allows us to be waiting, terminated, and running all at the same time.  Is this expected?
```
Comment (yastij) [inline pkg/apis/core/validation/validation.go]: I'm not sure there's a usecase where we'd want it

File: pkg/apis/core/validation/validation.go [hunk 2]
```diff
@@ -3570,9 +3586,22 @@ func ValidateContainerStateTransition(newStatuses, oldStatuses []core.ContainerS
...
 			}
+			if newStatus.State.Waiting != nil {
+				allErrs = append(allErrs, field.Forbidden(fldpath.Index(i).Child("state"), "may not go back to waiting"))
+			}
+			if newStatus.State.Running != nil {
+				allErrs = append(allErrs, field.Forbidden(fldpath.Index(i).Child("state"), "may not go back to running"))
+			}
+
+			// TODO this allows us to set Image, ImageID, and ContainerID back to empty after they had a value.  Is this intentional?  Should it accompanied by a state change?
+			// TODO this allows restartCount to move backwards.  Is this intentional?
```
Comment (yastij) [inline pkg/apis/core/validation/validation.go]: Not sure we want to move backwards

File: pkg/apis/core/validation/validation.go [hunk 3]
```diff
@@ -29,7 +29,7 @@ import (
 
 	"k8s.io/klog"
 
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
Comment (yastij) [inline pkg/apis/core/validation/validation.go]: nit: I think we can remove this change

**Review by yastij — COMMENTED** [review]
Generally lgtm

This isn't a high level validation, so I guess this is the right place

**Pull request closed without merging**